// tails-ui/lib/es6/src/components/ProgressBar.js
// Generated by BUCKLESCRIPT VERSION 2.2.0, PLEASE EDIT WITH CARE
'use strict';

import * as React from "react";
import * as ReasonReact from "reason-react/lib/es6/src/ReasonReact.js";

var component = ReasonReact.statelessComponent("ProgressBar");

function make($staropt$star, $staropt$star$1, _) {
  var color = $staropt$star ? $staropt$star[0] : "teal";
  var percent = $staropt$star$1 ? $staropt$star$1[0] : "50%";
  var newrecord = component.slice();
  newrecord[/* render */9] = (function () {
      return React.createElement("div", {
                  className: "w-full"
                }, React.createElement("div", {
                      className: "shadow w-full bg-grey-light rounded"
                    }, React.createElement("div", {
                          className: "bg-" + (String(color) + " rounded text-xs leading-none py-1 text-center text-white"),
                          style: {
                            width: percent
                          }
                        }, percent)));
    });
  return newrecord;
}

var $$default = ReasonReact.wrapReasonForJs(component, (function (jsProps) {
        return make(/* Some */[jsProps.color], /* Some */[jsProps.percent], /* array */[]);
      }));

export {
  component ,
  make ,
  $$default ,
  $$default as default,
  
}
/* component Not a pure module */
