// tails-ui/lib/es6/src/components/Select.js
// Generated by BUCKLESCRIPT VERSION 2.2.0, PLEASE EDIT WITH CARE
'use strict';

import * as React from "react";
import * as ReasonReact from "reason-react/lib/es6/src/ReasonReact.js";

var component = ReasonReact.statelessComponent("Select");

function make(label, value, name, placeholder, $staropt$star, children) {
  var color = $staropt$star ? $staropt$star[0] : "teal";
  var newrecord = component.slice();
  newrecord[/* render */9] = (function () {
      var labelComponent = label ? label[0] : null;
      var match = +(labelComponent === "");
      var tmp = {
        className: "bg-grey-lighter appearance-none border-2 border-grey-lighter hover:border-" + (String(color) + " rounded w-full py-2 px-4 mr-4 text-grey-darker")
      };
      if (name) {
        tmp.name = name[0];
      }
      if (placeholder) {
        tmp.placeholder = placeholder[0];
      }
      if (value) {
        tmp.value = value[0];
      }
      return React.createElement("div", {
                  className: "relative"
                }, match !== 0 ? React.createElement("label", {
                        className: "block uppercase tracking-wide text-grey-darker text-xs font-bold mb-2"
                      }, labelComponent) : null, React.createElement("select", tmp, children), React.createElement("div", {
                      className: "pointer-events-none absolute pin-y pin-r flex items-center px-2 text-grey-darker"
                    }, React.createElement("svg", {
                          className: "fill-current h-4 w-4",
                          viewBox: "0 0 20 20",
                          xmlns: "http://www.w3.org/2000/svg"
                        }, React.createElement("path", {
                              d: "M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"
                            }))));
    });
  return newrecord;
}

var $$default = ReasonReact.wrapReasonForJs(component, (function (jsProps) {
        return make(jsProps.label, jsProps.value, jsProps.name, jsProps.placeholder, /* Some */[jsProps.color], jsProps.children);
      }));

export {
  component ,
  make ,
  $$default ,
  $$default as default,
  
}
/* component Not a pure module */
